src/MessageFormatter.php

```diff
@@ -0,0 +1,180 @@
+    {
+        $this->timestampFormat = $timestampFormat;
+    }
+
+    /**
+     * Formats a log message for display as a string.
+     *
+     * The message structure follows that in {@see MessageCollection::$messages}.
+     *
+     * @param array $message The log message to be formatted.
+     *
+     * @throws LogRuntimeException for a callable "format" that does not return a string.
+     *
+     * @return string The formatted log message.
+     */
+    public function format(array $message): string
+    {
+        if ($this->format === null) {
+            return $this->defaultFormat($message);
+        }
+
+        $formatted = ($this->format)($message);
+
+        if (!is_string($formatted)) {
+            throw new LogRuntimeException(sprintf(
+                'The PHP callable "format" must returns a string, %s received.',
+                gettype($formatted)
+            ));
+        }
+
+        return $this->getPrefix($message) . $formatted;
+    }
+
+    /**
+     * Default formats a log message for display as a string.
+     *
+     * The message structure follows that in {@see MessageCollection::$messages}.
+     *
+     * @param array $message The log message to be default formatted.
+     *
+     * @return string The default formatted log message.
+     */
+    private function defaultFormat(array $message): string
+    {
+        [$level, $text, $context] = $message;
+        $timestamp = $context['time'] ?? microtime(true);
+        $category = $context['category'] ?? MessageCategoryFilter::DEFAULT;
+
+        $traces = [];
+        if (isset($context['trace']) && is_array($context['trace'])) {
+            foreach ($context['trace'] as $trace) {
+                if (isset($trace['file'], $trace['line'])) {
+                    $traces[] = "in {$trace['file']}:{$trace['line']}";
+                }
+            }
+        }
+
+        return $this->getTime($timestamp) . " {$this->getPrefix($message)}[{$level}][{$category}] {$text}"
+            . (empty($traces) ? '' : "\n    " . implode("\n    ", $traces));
+    }
+
+    /**
+     * Gets a string to be prefixed to the given message.
+     *
+     * If {@see MessageFormatter::$prefix} is configured it will return the result of the callback.
+     * The default implementation will return user IP, user ID and session ID as a prefix.
+     * The message structure follows that in {@see MessageCollection::$messages}.
+     *
+     * @param array $message The log message being exported.
+     *
+     * @throws LogRuntimeException for a callable "prefix" that does not return a string.
+     *
+     * @return string The log prefix string.
+     */
+    private function getPrefix(array $message): string
+    {
+        if ($this->prefix === null) {
+            return '';
+        }
+
+        $prefix = ($this->prefix)($message);
+
+        if (!is_string($prefix)) {
+            throw new LogRuntimeException(sprintf(
```

Comment:
Here we can use just `\RuntimeException`.

Reply:
Let's leave `LogRuntimeException`? This exception is thrown by the same condition in other classes of this package.